bnd-maven-plugin-parent/bnd-maven-plugin/src/main/java/org/bndtools/maven/BndMavenPlugin.java

```diff
@@ -0,0 +1,176 @@
+	private File manifestPath;
+	
+	@Parameter(defaultValue = "${project}", required = true, readonly = true)
+	private MavenProject project;
+	
+	@Parameter(property = "bnd.trace", defaultValue = "false", readonly = true)
+	private boolean trace;
+
+	public void execute() throws MojoExecutionException {
+		File bndFile = new File(project.getBasedir(), Project.BNDFILE);
+		
+		Builder builder = new Builder();
+		builder.setTrace(trace);
+		try {
+			if (bndFile.isFile())
+				builder.setProperties(bndFile);
+			
+			// Reject sub-bundle projects
+			Collection<? extends Builder> subs = builder.getSubBuilders();
+			if (subs.size() != 1)
+				throw new MojoExecutionException("Sub-bundles not permitted in a maven build");
+			File builderFile = builder.getPropertiesFile();
+			if (builderFile != null && !bndFile.equals(builderFile))
+				throw new MojoExecutionException("Sub-bundles not permitted in a maven build");
+
+			// Set bnd classpath
+			List<File> classpath = new LinkedList<File>();
+			@SuppressWarnings("unchecked")
+			Set<Artifact> artifacts = project.getArtifacts();
+			for (Artifact artifact : artifacts) {
+				File artifactFile = artifact.getFile();
+				if (artifactFile != null)
+					classpath.add(artifactFile);
+			}
+			classpath.add(classesDir);
+			builder.setClasspath(classpath.toArray(new File[classpath.size()]));
+
+			// Include local project packages automatically
+			String includes = builder.getProperty(Constants.INCLUDERESOURCE);
+			StringBuilder newIncludes = new StringBuilder().append(classesDir.getPath());
+			if (includes == null || includes.trim().length() == 0)
+				includes = newIncludes.toString();
+			else
+				includes = newIncludes.append(',').append(includes).toString();
+			builder.setProperty(Constants.INCLUDERESOURCE, includes);
+
+			// Set Bundle-Version
+			MavenVersion mvnVersion = new MavenVersion(project.getVersion());
+			builder.setProperty(Constants.BUNDLE_VERSION, mvnVersion.toBndVersion());
+
+			// Build bnd Jar (in memory)
+			Jar bndJar = builder.build();
+			
+			// Generate errors and warnings
+			List<String> warnings = builder.getWarnings();
+			if (warnings != null) for (String warning : warnings) {
+				getLog().warn(warning);
+			}
+			List<String> errors = builder.getErrors();
+			if (errors != null && !errors.isEmpty()) {
+				for (String error : errors) {
+					getLog().error(error);
+				}
+
+				if (errors.size() == 1)
+					throw new MojoExecutionException(errors.get(0));
+				else
+					throw new MojoExecutionException("Errors in bnd processing, see log for details.");
+			}
+
+			// Output manifest to <classes>/META-INF/MANIFEST.MF
+			Files.createDirectories(manifestPath.toPath().getParent());
+			FileOutputStream manifestOut = new FileOutputStream(manifestPath);
+			try {
+				bndJar.writeManifest(manifestOut);
+			} finally {
+				manifestOut.close();
+			}
+
+			// Expand Jar into target/classes
+			expandJar(bndJar, classesDir);
+
```

Comment:
Not sure if you need to check for errors again after writing the manifest. @pkriens told me to check again after writing the bundle since more errors could be detected. (For example: ${system} macro failure.). See <https://github.com/bndtools/bnd/blob/master/biz.aQute.bnd.gradle/src/aQute/bnd/gradle/BundleTaskConvention.groovy#L190>.